### ISSUE TITLE
Crash. UnsatisfiedLinkError Pixel 3a arm64-v8a

### ISSUE THREAD

**Issue by promanowicz** [realm/realm-kotlin#694]
Launching app with realm 0.9.0 fails on Pixel 3a 5G.
  The librealmc.so is in the apk file.
  
    java.lang.UnsatisfiedLinkError: dlopen failed: cannot locate symbol "__emutls_get_address" referenced by "/data/app/~~1OdRcdrRGBvzrkIJ8Ak36g==/com.....androidApp-GxBncVvTbA9_hL_OloW8kg==/base.apk!/lib/arm64-v8a/librealmc.so"...
        at java.lang.Runtime.loadLibrary0(Runtime.java:1077)
        at java.lang.Runtime.loadLibrary0(Runtime.java:998)
        at java.lang.System.loadLibrary(System.java:1656)
        at com.....compmulti.android.AndroidApplication.onCreate(AndroidApplication.kt:28)
        at com....Application.onCreate(CompanionApplication.kt:51)
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1211)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6725)
        at android.app.ActivityThread.access$1500(ActivityThread.java:247)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2053)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loopOnce(Looper.java:201)
        at android.os.Looper.loop(Looper.java:288)
        at android.app.ActivityThread.main(ActivityThread.java:7839)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)

**Comment by cmelchior** [comment]
Hi @promanowicz 

Can you post your build.gradle file? Specifically it is interesting to know which version of Android you are compiling for and what the device is running. 

Also, does this also crash if you run on the emulator or another device, or is it just this specific device?

**Comment by promanowicz** [comment]
Android 12, this also happens on Apple M1 emulator for Pixel 3a API 32.
The compile SDK was set to 31, update to 32 didn't fix the issue.
By default project was build via Intellij, but using AS does not help much.

```
buildscript {
    repositories {
        gradlePluginPortal()
        jcenter()
        google()
        mavenCentral()
    }
    dependencies {
        val kotlinVersion = "1.6.10"
        classpath(kotlin("gradle-plugin", version = kotlinVersion))
        classpath(kotlin("serialization", version = kotlinVersion))
        classpath("com.android.tools.build:gradle:7.0.4")
    }
}
```
```
import org.jetbrains.kotlin.gradle.plugin.mpp.KotlinNativeTarget

val ktor_version: String = "2.0.0-beta-1"

plugins {
    kotlin("multiplatform")
    kotlin("plugin.serialization")
    kotlin("native.cocoapods")
    id("com.android.library")
    id("io.realm.kotlin") version "0.9.0"
}
val realmVersion = "0.9.0"
val serialization_version = "1.3.2"
val kryptoVersion = "2.4.12"
kotlin {
    cocoapods {
        // Configure fields required by CocoaPods.
        summary = "Some description for a Kotlin/Native module"
        homepage = "Link to a Kotlin/Native module homepage"
        podfile = project.file("../iosApp/Podfile")
        framework {
            baseName = "shared"
            isStatic = false
        }
    }
    android()
//    ios()
    iosX64()
    iosArm64()
    iosSimulatorArm64()
    sourceSets {
        val commonMain by getting {
            dependencies {
                implementation ("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.6.0-native-mt"){
                    version {
                        strictly("1.6.0-native-mt")
                    }
                }
                implementation("io.ktor:ktor-client-core:$ktor_version")
                implementation("org.jetbrains.kotlin:kotlin-reflect:1.6.10")
                implementation("co.touchlab:stately-common:1.2.0")
                implementation("io.realm.kotlin:library-base:${realmVersion}")
                implementation("org.jetbrains.kotlinx:kotlinx-serialization-json:$serialization_version")
                implementation("org.jetbrains.kotlinx:kotlinx-datetime:0.3.1")
                implementation("com.soywiz.korlibs.krypto:krypto:$kryptoVersion")
            }
        }
        val commonTest by getting {
            dependencies {
                implementation ("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.6.0-native-mt"){
                    version {
                        strictly("1.6.0-native-mt")
                    }
                }
                implementation(kotlin("test"))
                implementation("io.kotest:kotest-assertions-core:5.0.1")
            }
        }
        val androidMain by getting {
            dependencies {
                implementation("com.google.android.material:material:1.5.0")
                implementation("io.ktor:ktor-client-android:$ktor_version")
                implementation("com.jakewharton:process-phoenix:2.1.2")
            }
        }
        val androidTest by getting {
            dependencies {
                implementation("junit:junit:4.13.2")
            }
        }
        val iosX64Main by getting
        val iosArm64Main by getting
        val iosSimulatorArm64Main by getting
        val iosMain by creating {
            dependsOn(commonMain)
            iosX64Main.dependsOn(this)
            iosArm64Main.dependsOn(this)
            iosSimulatorArm64Main.dependsOn(this)
            dependencies {
                implementation("io.ktor:ktor-client-ios:$ktor_version")
            }
        }
        val iosX64Test by getting
        val iosArm64Test by getting
        val iosSimulatorArm64Test by getting
        val iosTest by creating {
            dependsOn(commonTest)
            iosX64Test.dependsOn(this)
            iosArm64Test.dependsOn(this)
            iosSimulatorArm64Test.dependsOn(this)
        }
    }
}

android {
    compileSdk = 32
    sourceSets["main"].manifest.srcFile("src/androidMain/AndroidManifest.xml")
    defaultConfig {
        minSdk = 24
        targetSdk = 29
    }
}
```
additionaly the MMP Android project is imported to another Android project
where 
    ```
packagingOptions {
        pickFirst("**/libc++_shared.so")
    }
```
is added.

**Comment by cmelchior** [comment]
Any reason you had to add

```
packagingOptions {
pickFirst("**/libc++_shared.so")
}
```

I strongly suspect this is what is causing the issue?

**Comment by promanowicz** [comment]
Yes, and yes. 
React Native is dependency to the project and it looks some other sdk's
have had similar issues. 
Eg: that https://github.com/facebook/react-native/issues/30459
Maybe RN is build with different NDK - 21.4.7075529 and thats causing the issue.
https://github.com/facebook/react-native/wiki/Building-from-source
I tried to force usage local libc++.so by placing it in src/main/jniLibs but this results in RN crash.

**Comment by cmelchior** [comment]
Yes, we are bundling the `libc++_shared` library right now and that is the most likely root cause. It is also described here: https://developer.android.com/ndk/guides/cpp-support#one_stl_per_app

We could potentially statically link it into our SDK, but we would need to evaluate it first.

Unfortunately, right now, it looks like Realm Kotlin and React Native are incompatible. 

I do wonder why React Native are still using NDK 21 though 🤔

**Comment by cortinico** [comment]
> I do wonder why React Native are still using NDK 21 though 🤔

Chiming here. React Native uses the side-by-side NDK default version provided by AGP (which for 7.1.x is r21): https://developer.android.com/studio/releases/gradle-plugin?buildsystem=cmake

Also @promanowicz, what's the reason why you're using realm-kotlin rather than realm-js?

**Comment by cmelchior** [comment]
This is being rolled into #710. We will need to evaluate exactly which approach to take here.

**Comment by nhachicha** [comment]
Hi @promanowicz this is fixed now in `0.10.1` could you try, please.
